### ISSUE TITLE
Integrate lomse in JUCE

### ISSUE THREAD

**Issue by ffAudio** [lenmus/lomse#282]
Thanks for this awesome library!

I try to use this library in JUCE and I copied from the approach of the QT example.
Unfortunately it crashes somewhere in the lomse library.

This is the crash dump:
```
Crashed Thread:        0  JUCE Message Thread  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [1768]

VM Regions Near 0:
--> 
    __TEXT                 000000010ca77000-000000010d260000 [ 8100K] r-x/r-x SM=COW  /Users/USER/*/demo.app/Contents/MacOS/demo

Thread 0 Crashed:: JUCE Message Thread  Dispatch queue: com.apple.main-thread
0   liblomse.0.27.0.so            	0x000000010eace0b8 lomse::ScreenDrawer::reset(agg::row_accessor<unsigned char>&, agg::rgba8T<agg::linear>) + 40
1   liblomse.0.27.0.so            	0x000000010e93b0ac lomse::GraphicView::draw_graphic_model() + 204
2   liblomse.0.27.0.so            	0x000000010e93af92 lomse::GraphicView::draw_all() + 66
3   liblomse.0.27.0.so            	0x000000010e93ab79 lomse::GraphicView::redraw_bitmap() + 41
4   liblomse.0.27.0.so            	0x000000010e94c321 lomse::Interactor::redraw_bitmap() + 273
5   com.FoleysFinestAudio.demo	0x000000010ca7b934 MainComponent::paint(juce::Graphics&) + 404 (MainComponent.cpp:80)
6   com.FoleysFinestAudio.demo	0x000000010cecc219 juce::Component::paintComponentAndChildren(juce::Graphics&) + 297 (juce_Component.cpp:1926)
7   com.FoleysFinestAudio.demo	0x000000010cecc0b7 juce::Component::paintEntireComponent(juce::Graphics&, bool) + 1047
8   com.FoleysFinestAudio.demo	0x000000010cecbc63 juce::Component::paintWithinParentContext(juce::Graphics&) + 147 (juce_Component.cpp:1911)
9   com.FoleysFinestAudio.demo	0x000000010cecc66b juce::Component::paintComponentAndChildren(juce::Graphics&) + 1403 (juce_Component.cpp:1968)
10  com.FoleysFinestAudio.demo	0x000000010cecc0b7 juce::Component::paintEntireComponent(juce::Graphics&, bool) + 1047
11  com.FoleysFinestAudio.demo	0x000000010cf971ad juce::ComponentPeer::handlePaint(juce::LowLevelGraphicsContext&) + 621 (juce_ComponentPeer.cpp:132)
12  com.FoleysFinestAudio.demo	0x000000010d03e9d8 juce::NSViewComponentPeer::invokePaint(juce::LowLevelGraphicsContext&) + 40 (juce_mac_NSViewComponentPeer.mm:992)
13  com.FoleysFinestAudio.demo	0x000000010d03e4a1 juce::NSViewComponentPeer::drawRect(CGContext*, CGRect, float) + 225 (juce_mac_NSViewComponentPeer.mm:894)
14  com.FoleysFinestAudio.demo	0x000000010d03e3ac juce::NSViewComponentPeer::drawRect(CGRect) + 428 (juce_mac_NSViewComponentPeer.mm:886)
15  com.FoleysFinestAudio.demo	0x000000010d03be92 juce::JuceNSViewClass::drawRect(objc_object*, objc_selector*, CGRect) + 130 (juce_mac_NSViewComponentPeer.mm:1747)
16  com.apple.AppKit              	0x00007fff3582121e _NSViewDrawRect + 66
17  com.apple.AppKit              	0x00007fff3581faac -[NSView(NSInternal) _recursive:displayRectIgnoringOpacity:inContext:shouldChangeFontReferenceColor:stopAtLayerBackedViews:] + 1627
18  com.apple.AppKit              	0x00007fff3581f43f __46-[NSView(NSLayerKitGlue) drawLayer:inContext:]_block_invoke + 192
19  com.apple.AppKit              	0x00007fff3581f19c -[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:] + 1720
20  com.apple.QuartzCore          	0x00007fff42b9fe02 CABackingStoreUpdate_ + 577
21  com.apple.QuartzCore          	0x00007fff42c01a0d invocation function for block in CA::Layer::display_() + 53
22  com.apple.QuartzCore          	0x00007fff42b9f020 -[CALayer _display] + 1830
23  com.apple.AppKit              	0x00007fff3581e74e _NSBackingLayerDisplay + 528
24  com.apple.AppKit              	0x00007fff35802c4f -[_NSViewBackingLayer display] + 811
25  com.apple.QuartzCore          	0x00007fff42b9e551 CA::Layer::display_if_needed(CA::Transaction*) + 627
26  com.apple.QuartzCore          	0x00007fff42b8c7c6 CA::Context::commit_transaction(CA::Transaction*) + 342
27  com.apple.QuartzCore          	0x00007fff42b8bea6 CA::Transaction::commit() + 596
28  com.apple.AppKit              	0x00007fff357f984d __65+[CATransaction(NSCATransaction) NS_setFlushesWithDisplayRefresh]_block_invoke + 274
29  com.apple.CoreFoundation      	0x00007fff381605b8 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
30  com.apple.CoreFoundation      	0x00007fff381604ed __CFRunLoopDoObservers + 451
31  com.apple.CoreFoundation      	0x00007fff38102b44 __CFRunLoopRun + 1136
32  com.apple.CoreFoundation      	0x00007fff38102482 CFRunLoopRunSpecific + 455
33  com.apple.HIToolbox           	0x00007fff373611ab RunCurrentEventLoopInMode + 292
34  com.apple.HIToolbox           	0x00007fff37360ded ReceiveNextEventCommon + 355
35  com.apple.HIToolbox           	0x00007fff37360c76 _BlockUntilNextEventMatchingListInModeWithFilter + 64
36  com.apple.AppKit              	0x00007fff356f877d _DPSNextEvent + 1135
37  com.apple.AppKit              	0x00007fff356f746b -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1361
38  com.apple.AppKit              	0x00007fff356f1588 -[NSApplication run] + 699
39  com.FoleysFinestAudio.demo	0x000000010cd9ceb2 juce::MessageManager::runDispatchLoop() + 162 (juce_mac_MessageManager.mm:360)
40  com.FoleysFinestAudio.demo	0x000000010cd9cd87 juce::JUCEApplicationBase::main() + 391 (juce_ApplicationBase.cpp:262)
41  com.FoleysFinestAudio.demo	0x000000010cd9cb9c juce::JUCEApplicationBase::main(int, char const**) + 60 (juce_ApplicationBase.cpp:240)
42  com.FoleysFinestAudio.demo	0x000000010ca78703 main + 51 (Main.cpp:101)
43  libdyld.dylib                 	0x00007fff640bf3d5 start + 1
```

I'll attach my code here (hope you don't mind). Once it works feel free to put it in your project :-)

Just replace the MainComponent.cpp and .h in a default AudioApplication in JUCE (I'm happy to provide info how to do that).

```
// MainComponent.h
#pragma once

#include <JuceHeader.h>

//lomse headers
#include <lomse_doorway.h>
#include <lomse_document.h>

//==============================================================================
/*
    This component lives inside our window, and this is where you should put all
    your controls and content.
*/
class MainComponent  : public juce::AudioAppComponent
{
public:
    //==============================================================================
    MainComponent();
    ~MainComponent() override;

    //==============================================================================
    void prepareToPlay (int samplesPerBlockExpected, double sampleRate) override;
    void getNextAudioBlock (const juce::AudioSourceChannelInfo& bufferToFill) override;
    void releaseResources() override;

    //==============================================================================
    void paint (juce::Graphics& g) override;
    void resized() override;

private:
    void createDummyScore();

    //==============================================================================
    juce::Image                       buffer;

    lomse::LomseDoorway               lomse;      //the Lomse library doorway
    std::unique_ptr<lomse::Presenter> presenter;  //relates the View, the Document and the Interactor

    JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (MainComponent)
};

// MainComponent.cpp
#include "MainComponent.h"

//lomse headers
#include <lomse_graphic_view.h>
#include <lomse_interactor.h>
#include <lomse_presenter.h>


//==============================================================================
MainComponent::MainComponent()
{
    //initialize the Lomse library
    auto format = lomse::k_pix_format_abgr32;
    lomse.init_library (format, 150, false);

    // calls resized()
    setSize (800, 600);

    // Some platforms require permissions to open input channels so request that here
    if (juce::RuntimePermissions::isRequired (juce::RuntimePermissions::recordAudio)
        && ! juce::RuntimePermissions::isGranted (juce::RuntimePermissions::recordAudio))
    {
        juce::RuntimePermissions::request (juce::RuntimePermissions::recordAudio,
                                           [&] (bool granted) { setAudioChannels (granted ? 2 : 0, 2); });
    }
    else
    {
        // Specify the number of input and output channels that we want to open
        setAudioChannels (2, 2);
    }

    createDummyScore();
}

MainComponent::~MainComponent()
{
    // This shuts down the audio device and clears the audio source.
    shutdownAudio();
}

void MainComponent::createDummyScore()
{
    presenter.reset (lomse.new_document (lomse::k_view_vertical_book,
                                         "(lenmusdoc (vers 0.0)"
                                         "(content "
                                         "(para (txt \"Hello world!\"))"
                                         "(score (vers 2.0) "
                                         "(instrument (musicData (clef G)(key C)(time 2 4)(n c4 q) )))"
                                         ")"
                                         ")",
                                         lomse::Document::k_format_ldp));
}

//==============================================================================
void MainComponent::prepareToPlay (int samplesPerBlockExpected, double sampleRate) {}

void MainComponent::getNextAudioBlock (const juce::AudioSourceChannelInfo& bufferToFill)
{
    bufferToFill.clearActiveBufferRegion();
}

void MainComponent::releaseResources() {}

//==============================================================================
void MainComponent::paint (juce::Graphics& g)
{
    // (Our component is opaque, so we must completely fill the background with a solid colour)
    g.fillAll (getLookAndFeel().findColour (juce::ResizableWindow::backgroundColourId));

    //get the pointer to the interactor, set the rendering buffer and
    //register for receiving desired events
    if (auto interactor = presenter->get_interactor(0).lock())
    {
        buffer.clear (buffer.getBounds(), juce::Colours::darkgrey);

        //connect the View with the window buffer
        juce::Image::BitmapData data (buffer, 0, 0, buffer.getWidth(), buffer.getHeight(), juce::Image::BitmapData::writeOnly);

        interactor->set_rendering_buffer ((lomse::RenderingBuffer*) data.data);
        interactor->redraw_bitmap();
    }

    // You can add your drawing code here!
    g.drawImageAt (buffer, 10, 10);
}

void MainComponent::resized()
{
    // This is called when the MainContentComponent is resized.
    // If you add any child components, this is where you should
    // update their positions.
    buffer = juce::Image (juce::Image::ARGB, getWidth()-20, getHeight()-20, false);
}
```

Thank you for any hint how to use it right,
Daniel

**Comment by cecilios** [comment]
### Using Lomse in JUCE

I have no experience using JUCE and have never tested Lomse in JUCE. But Lomse has been used with JUCE (e.g. https://github.com/hugbug/conpianist)  and I have seen it is used as follows (in line with Qt sample):

```
Application framework:    JUCE
Operating system:    any


Lomse library initialization
---------------------------------------
probably
    //Resolution: 96 * UI scale
	m_scale = m_settings.zoomUi * Desktop::getInstance().getDisplays().getMainDisplay().scale;
	int resolution = int(96 * m_scale);
	
	//pixel format
	int pixel_format = k_pix_format_rgba32;

	//Lomse default y axis direction is 0 coordinate at top and increases
	//downwards. For JUCE the Lomse default behaviour is the right behaviour.
	bool reverse_y_axis = false;

	//initialize the Lomse library with these values
	m_lomse.init_library(pixel_format, resolution, reverse_y_axis);


Creating the rendering buffer
------------------------------

    //Memory for the bitmap:    juce::Image
    //Bitmap format:            juce::Image::PixelFormat::ARGB

	m_docScale = imoDoc->get_page_content_scale(); //scale is computed by lomse when rebuilding the GraphicModel

	//create image to fit the whole page
	m_image.reset(new juce::Image(juce::Image::PixelFormat::ARGB,
		int(width / m_docScale), int(height / m_docScale), false, SoftwareImageType()));
		
	//creates a bitmap of specified size and associates it to the rendering
	//buffer for the view. Any existing buffer is automatically deleted
	juce::Image::BitmapData bitmap(*m_image, juce::Image::BitmapData::readWrite);
	m_rbuf_window.attach(bitmap.data, m_image->getWidth(), m_image->getHeight(), bitmap.lineStride);


Paint event
-----------------

void MyWindow::paint(Graphics& g)
{
	if (m_presenter && m_image)
	{
		g.drawImage(*m_image, 0, 0, getWidth(), getHeight(), 0, 0, m_image->getWidth(), m_image->getHeight());
	}
	else
	{
		String text = "Some error message";
		juce::Rectangle<int> rec(20, 80, getWidth() - 40, getHeight() - 100);
		g.drawFittedText(text, rec, Justification::centredTop, 100, 1);
	}
}
```

### Crash cause

The trace points to some problem (null object) when rendering the music. In your code I have observed that you do not set a rendering buffer when creating a document (createDummyScore) but when the paint event is processed. This causes two problems:
- Lomse send paint events **after** having rendered the score. Thus, when the dummy document is processed by Lomse, it tries to render it on a null buffer and this is the cause of the crash.
- In your app `paint` method, after having displayed the received buffer, you creates a new buffer. This is not needed as the old buffer is valid while the window size doe not change. Setting a new buffer here without being necessary is more memory management workload, and forces Lomse to repaint all again every time you allocates a new buffer. This is inefficient.

I would suggets to follow the same approach than in Qt sample. Hope this helps!

**Comment by cecilios** [comment]
I have also observed that you uses lomse::k_pix_format_abgr32 and suggested usage is to use k_pix_format_rgba32. Lomse should work with abgr32 but if after fixing the rendering buffer issue you experience problems with the image format (bad colors) try with k_pix_format_rgba32.

**Comment by ffAudio** [comment]
Thank you for your quick reply. I am still trying to sort things.

The Image is persistent in my code as well. The Image::BitmapData is an access helper, because juce::Image might use platform dependent image classes (like NSImage on mac). I see that AGG has the same abstraction again, but that shouldn't do no harm.

When you say it sends paint events after having rendered the score, do you mean there are asynchronous calls going on in a separate thread? But the debugger stops at `interactor->redraw_bitmap();`, which is where the Image::Bitmap is well alive.

Can you elaborate on the line 
```
m_rbuf_window.attach(bitmap.data, m_image->getWidth(), m_image->getHeight(), bitmap.lineStride);
```
?
What type is m_rbuf_window? It doesn't seem to be a lomse function, at least I couldn't spot it in the lomse API doxygen.

I also realised that the reinterpret cast of data* to `lomse::RenderingBuffer` might not be a good idea, since that is a struct.
I tried then to setup that struct with the arguments you used in that attach method, since there is a constructor that seems to fit here. But it didn't change the behaviour.
```
    if (auto interactor = presenter->get_interactor(0).lock())
    {
        image.clear (image.getBounds(), juce::Colours::darkgrey);

        //connect the View with the window buffer
        juce::Image::BitmapData data (image, 0, 0, image.getWidth(), image.getHeight(), juce::Image::BitmapData::writeOnly);
        lomse::RenderingBuffer buffer (data.data, image.getWidth(), image.getHeight(), 0);

        interactor->set_rendering_buffer (&buffer);
        interactor->redraw_bitmap();
    }
```

Thanks for the hint about the pixel format, I tried different ones, this one worked when I used FFMPEG with juce::Image, but I will try the different ones, once I get over that crash.

Thanks again.

**Comment by cecilios** [comment]
### The RenderingBuffer object

Let's start by the last question, the `RenderingBuffer`. Lomse renders the document on a bitmap provided by your application. The pointer to this bitmap, as well as the bitmap size (width, height) and stride (bytes per row) are passed to Lomse in an instance of class `RenderingBuffer`. Your application needs to allocate an object of this class, with at least the same scope than the `LomseDoorway` object. They are both normally defined in the same place, In the tutorials it was named `m_rbuf_window`. Just define it at the same place that you define lomse doorway, probably:
```
class MainComponent
{
protected:
    LomseDoorway&       lomse;
    RenderingBuffer     m_rbuf_window;
    ...
```

When a new Lomse::View is created (because your app. asks lomse to open a document or to create a new one) your app. has to inform lomse about the rendering buffer to use for this document. Also, normally, event handlers are defined at this point,:
```
void MainComponent::createDummyScore()
{
    presenter.reset (lomse.new_document (lomse::k_view_vertical_book,
                                         "(lenmusdoc (vers 0.0)"
                                         "(content "
                                         "(para (txt \"Hello world!\"))"
                                         "(score (vers 2.0) "
                                         "(instrument (musicData (clef G)(key C)(time 2 4)(n c4 q) )))"
                                         ")"
                                         ")",
                                         lomse::Document::k_format_ldp));

    //connect the View with the window buffer
    interactor->set_rendering_buffer(&m_rbuf_window);  //probably not yet initialized, doesn't matter

    //register to receive desired events (optional)
    interactor->add_event_handler(k_update_window_event, this, wrapperUpdateWindow);
    interactor->add_event_handler(k_do_play_score_event, this, wrapperPlayScore);
    interactor->add_event_handler(k_pause_score_event, this, wrapperPlayScore);
    interactor->add_event_handler(k_stop_playback_event, this, wrapperPlayScore);
        ... other event handlers that could be needed by your app.
}
```

Notice that at this point `m_rbuf_window` could be un-initialized. But it doesn't matter. You are just providing Lomse a ptr to the `RenderingBuffer` object. This object has to be properly initialized before asking Lomse to render anything.  As said, this **is not the Image buffer** but a `RenderingBuffer` object containing a pointer to the image buffer. Initialization is done when you allocate a buffer, e.g. in the size event (please, review, I know nothing of juce):
```
void MainComponent::resized()
{
    // This is called when the MainContentComponent is resized.
    // If you add any child components, this is where you should
    // update their positions.

    int width = getWidth()-20;
    int height = getHeight()-20;
    buffer = juce::Image (juce::Image::ARGB, width, height, false);
    juce::Image::BitmapData data (buffer, 0, 0, width, height, juce::Image::BitmapData::writeOnly);

    //determine stride
    const int BYTES_PER_PIXEL = 4;   //the chosen format is ARGB, 32 bits -> 4 bytes
    int stride = width * BYTES_PER_PIXEL;     //number of bytes per row

    //Plug-in this new info to the RenderinBuffer object
    m_rbuf_window.attach(data.data, width, height, stride);
}
```

### Paint events

> When you say it sends paint events after having rendered the score, do you mean there are asynchronous calls going on in a separate thread?

Sorry! In my previous answer I wrote something that is wrong: Lomse will not trigger any paint event in this case. 

Lomse only uses a separate thread for playback. It will trigger `lomse paint events` during playback to inform your application that the screen needs to be updated (the rendering buffer has been already updated and your app. only has to update the screen). This is because as playback goes on, visual effects (the cursor on the current beat, the notes being played or a page advance) are updated and it is necessary to repaint the window.

These `lomse paint events` are totally synchronous and take place when Lomse updates the image during playback and informs your app. by invoking the callback that your application should have set for receiving Lomse events, and not by generating system paint events.


### The crash in your app.

> I also realized that the reinterpret cast of data* to lomse::RenderingBuffer might not be a good idea

Yes, that's wrong, you are not passing a `RenderingBuffer` object but the bitmap memory cast as to `RenderingBuffer`. Thus, values for width, height and stride, as well as the ptr. to the bitmap, are probably 0.

As to your new code, the scope of the `RenderingBuffer` is local. But as you allocate a new buffer for every repaint probably this is not the cause of the crash. Stride is set to 0, and probably this is the cause of the crash with this updated code, but not sure.

**Comment by cecilios** [comment]
To be more accurate:

- `RenderingBuffer` is a wrapper class containing the pointer to the real buffer and some information. Its lifespan should be at least the same lifespan than memory used for rendering the score (the image buffer).

- A new buffer has to be allocated whenever the window is resized, and the `RenderingBuffer` object must be updated with the new information.

- Lomse will render the score on demand on the passed buffer, that is, it is your application who decides when to render. Therefore, it is feasible to allocate a new buffer for each paint event, but this is very inefficient as your app. will be constantly allocating and de-allocating a big amount of memory space.

- During playback it is Lomse who takes the initiative to render the score (highlight effects, tempo line, score auto-scroll to maintain the action on the displayed score chunk) and generates events, such as `EventPaint` or `EventUpdateViewport` to inform your app. of the need to update the window. Thus, if your application uses the Lomse playback facilities, the lifespan of the image buffer and the `RenderingBuffer` object should be at least all the playback time.

- The recommended approach is to allocate a new buffer when the window is resized. The lifespan of this buffer should be until next window resize.


This issue of the `RenderingBuffer` object is a symptom of a bad design in Lomse API. I'm realizing that `RenderingBuffer` should be a Lomse internal object, and any application using Lomse should just pass a ptr. to the image buffer and its size. I'm opening issue #283 to study this in depth.

**Comment by ffAudio** [comment]
That's great, although keeping the RenderingBuffer as member variable didn't avoid the crash. I am still investigating.

Meanwhile I looked a bit deeper into the rendering of lomse and found the lomse::Drawer interface. Do you think it is doable to write a `class JuceDrawer : public lomse::Drawer` that draws the primitives on the juce::Graphics context?
It has two advantages, it avoids the bitmap memory, might be faster and avoids using the AGG, of which I am suspicious about anyway, given that it seems to be GPL?

**Comment by ffAudio** [comment]
BTW about allocating the buffer, like I said, in juce the juce::Image holds the image data. It is implemented as a shared data to avoid copying when passing it around.
Creating the juce::Image::PixelData will not allocate but only create a translation, potentially adding the conversion of different pixel formats and so on (but since we access the data directly, we still need to pick the right format ourselves).
If you are curious, here is the API doc: https://docs.juce.com/develop/classImage_1_1BitmapData.html

Thank you for your help.

**Comment by ffAudio** [comment]
Just found the culprit:

In the RendererFactory, not all pix_formats return a valid renderer. I guess adding an `assert (false)` in the `default: return nullptr;` would help to find that problem sooner :-)

Thanks for all your help, would never found it without your explanations!

**Comment by cecilios** [comment]
Glad that the problem is solved!

> Just found the culprit:

Aggg! I forget that although the initial idea was to support all pixel formats supported by AGG, I only implemented some of of them! Sorry! I will review the code to inform if trying to use a format not yet implemented. Thank you!

As to your other points:

> Do you think it is doable to write a class JuceDrawer : public lomse::Drawer that draws the primitives on the juce::Graphics context?

Should be totally feasible and simple, as Lomse was designed with that idea on mind. Unfortunately, reviewing the code, I have noticed that all member functions accept a parameter of class `SreenDrawer*`, but no problem, it is just a careful global change to `Drawer*` that, in any case, should be done.

The second point to change is in the only point in which the `Drawer` instance is created, in method (`lomse_injectors.cpp`):
```
View* Injector::inject_View(LibraryScope& libraryScope, int viewType,
                            Document* UNUSED(pDoc))
{
    ScreenDrawer* pDrawer = Injector::inject_ScreenDrawer(libraryScope);
    return ViewFactory::create_view(libraryScope, viewType, pDrawer);
}
```

As you see, now it always injects a `ScreenDrawer`. As this is an internal method, there is no problem in changing it. I would remove the unused parameter and add a `drawer type` parameter. But this parameter should be propagated, to arrive to LomseDoorway methods, that originates the invokation chain:
```
inject_Presenter()
PresenterBuilder::new_document()
PresenterBuilder::open_document()
LomseDoorway::new_document()
LomseDoorway::open_document()
```

`LoomseDoorway` methods are public and adding an additional parameter will break backwards compatibility. And the drawer to use is a global parameter, so instead of adding the new parameter to `LomseDoorway` document creation methods, this new parameter can be added to the `LomseDoorway::init_library` method. This also breaks backwards compatibility but is is acceptable and, in any case, an alternative new method with different name can be used.

I hope that these changes will be all that is needed. With them, the library should be totally ready to accept other drawers, JUCE, SVG stream or whatever. Even if you decide not to write the JuceDrawer, these changes should be done in any case. I add this to the TODO list. A PR is welcome !  ;)

> ... using the AGG, of which I am suspicious about anyway, given that it seems to be GPL?

No, I started to use AGG around 2004 and until release 2.4 included, AGG was released with MIT license. Later (~2006), his developer, Maxim Shemanarev, started to work for a Swiss company and decided to change the license to GPL, so he announced the license change for the new versions and released the same code as release 2.5 with GPL license. Unfortunately, Maxim died in 2013 in an accident. AGG was appreciated and used in a lot of projects, and so when the change to GLP took place, a fork of the 2.4 code with MIT license was created and it is since them maintained at SourceForge. But is is such a good code that no patches are needed! AGG is used in many projects such as the Haiku operating system, Matplotlib or OpenStreetMap. Don't worry, it is safe to use AGG.

> Thanks for all your help, would never found it without your explanations!

Thanks to you. The problems found using Lomse, questions, comments and suggestions, are very important to improve the library. Thank you!
